pkg/controller/service/service_controller.go

```diff
@@ -741,64 +661,94 @@ func (s *ServiceController) syncService(key string) error {
-			// were added into the queue during the delay, but it does not mess as when handling the retry,
-			// it always get the last service info from service store
-			s.workingQueue.AddAfter(obj, delay)
-		}(key, retryDelay)
-	} else if err != nil {
-		runtime.HandleError(fmt.Errorf("Failed to process service. Not retrying: %v", err))
 	}
-	return nil
+
+	if s.needsDeletion(service) {
+		return s.processServiceDeletion(service, key)
+	}
+
+	return s.processServiceUpdate(service, key)
 }
 
 // Returns an error if processing the service deletion failed, along with a time.Duration
 // indicating whether processing should be retried; zero means no-retry; otherwise
 // we should retry after that Duration.
-func (s *ServiceController) processServiceDeletion(key string) (error, time.Duration) {
-	cachedService, ok := s.cache.get(key)
-	if !ok {
-		return fmt.Errorf("Service %s not in cache even though the watcher thought it was. Ignoring the deletion.", key), doNotRetry
-	}
-	return s.processLoadBalancerDelete(cachedService, key)
+func (s *ServiceController) processServiceDeletion(service *v1.Service, key string) error {
+	return s.processLoadBalancerDelete(service, key)
 }
 
-func (s *ServiceController) processLoadBalancerDelete(cachedService *cachedService, key string) (error, time.Duration) {
-	service := cachedService.state
-	// delete load balancer info only if the service type is LoadBalancer
-	if !wantsLoadBalancer(service) {
-		return nil, doNotRetry
-	}
+func (s *ServiceController) processLoadBalancerDelete(service *v1.Service, key string) error {
 	s.eventRecorder.Event(service, v1.EventTypeNormal, "DeletingLoadBalancer", "Deleting load balancer")
 	err := s.balancer.EnsureLoadBalancerDeleted(s.clusterName, service)
 	if err != nil {
 		message := "Error deleting load balancer (will retry): " + err.Error()
 		s.eventRecorder.Event(service, v1.EventTypeWarning, "DeletingLoadBalancerFailed", message)
-		return err, cachedService.nextRetryDelay()
+		return err
 	}
 	s.eventRecorder.Event(service, v1.EventTypeNormal, "DeletedLoadBalancer", "Deleted load balancer")
-	s.cache.delete(key)
 
-	cachedService.resetRetryDelay()
-	return nil, doNotRetry
+	// Now that we've successfully cleaned up the load balancer resources we can remove the finalizer
+	// and allow the service to be deleted.
+	updatedFinalizers := removeFinalizer(service)
+	if len(updatedFinalizers) != len(service.Finalizers) {
+		updated := service.DeepCopy()
+		updated.Finalizers = updatedFinalizers
+		err = s.patchService(service, updated)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// TODO(horwitz): test me
+func removeFinalizer(service *v1.Service) []string {
+	finalizers := []string{}
+	for _, finalizer := range service.Finalizers {
+		if finalizer != finalizerName {
+			finalizers = append(finalizers, finalizer)
+		}
+	}
+	return finalizers
+}
+
+// TODO(horwitz): test me
+func addFinalizer(service *v1.Service) []string {
```

Comment:
https://github.com/kubernetes/kubernetes/blob/9807360fe30d2947eea5833aeaf19d47c9c3c89c/federation/pkg/federation-controller/util/finalizers/finalizers.go might be of interest.

Reply:
I didn't want to use that since it's under the federation controller and i'd need to refactor it out.

Reply:
It might be worth factoring it out into apimachinery. If you think there would likely be support for doing so I'd be happy to do the leg work.